### PR TITLE
refactor(backend): migrate changelogs service from Thrift to Spring Boot REST

### PR DESCRIPTION
**Pull request by Shivamrut**
### Summary
Migrate the changelogs backend service from Thrift RPC to Spring Boot REST, following the established pattern from health, configurations, and search migrations. This PR also introduces a `PaginatedResult<T>` wrapper DTO to solve JSON serialization of `Map<PaginationData, List<T>>` across the REST boundary, and fixes a `ClassCastException` in the resource-server's pagination pipeline caused by Thrift-typed comparators applied to POJO objects.

### Layer 1 — Changelogs POJOs (libraries/service-api)

| File | Action | Purpose |
|------|--------|---------|
| `services/changelogs/ChangeLogs.java` | new | POJO replacing Thrift `ChangeLogs` struct. Fields: `documentId`, `documentType`, `operation`, `userEdited`, `changeTimestamp`, `changes` (Set), `referenceDoc` (Set), `info` (Map), etc. Constructor sets `type = "changeLogs"` |
| `services/changelogs/ChangedFields.java` | new | POJO replacing Thrift `ChangedFields` struct. Fields: `fieldName`, `fieldValueOld`, `fieldValueNew` |
| `services/changelogs/ReferenceDocData.java` | new | POJO replacing Thrift `ReferenceDocData` struct. Fields: `refDocId`, `dbName`, `refDocType`, `refDocOperation` |
| `services/changelogs/Operation.java` | new | Enum replacing Thrift `Operation`. 21 values: `CREATE`, `UPDATE`, `DELETE`, `MERGE_COMPONENT`, `OBLIGATION_UPDATE`, etc. |
| `services/changelogs/ChangelogSortColumn.java` | new | Enum for sort column selection in pagination. Values: `BY_CHANGE_TIMESTAMP`, `BY_USER_EDITED` |
| `services/common/PaginatedResult.java` | new | Generic wrapper DTO with `paginationData` and `data` fields. Solves the `Map<PaginationData, ...>` JSON key serialization problem (see "Design Decisions" below) |

All POJOs use `@Data` + `@Accessors(chain = true)` from Lombok, `@JsonInclude(NON_NULL)` from Jackson.

### Layer 2 — Converter Additions (backend/common)

| File | Action | Purpose |
|------|--------|---------|
| `common/converter/ThriftConverter.java` | modified | Added 6 new converter methods for changelogs types, plus 2 missing POJO imports |

New converter methods:
- `fromThriftChangeLogs()` — converts Thrift `ChangeLogs` → POJO `ChangeLogs` with stream-mapped nested sets
- `fromThriftChangedFields()` — converts Thrift `ChangedFields` → POJO `ChangedFields`
- `fromThriftReferenceDocData()` — converts Thrift `ReferenceDocData` → POJO `ReferenceDocData`
- `fromThriftOperation()` — converts Thrift `Operation` enum → POJO `Operation` enum
- `toThriftPaginationData()` — converts POJO `PaginationData` → Thrift `PaginationData`
- `fromThriftPaginationData()` — converts Thrift `PaginationData` → POJO `PaginationData`

### Layer 3 — Backend Service (backend/changelogs)

| File | Action | Purpose |
|------|--------|---------|
| `backend/changelogs/pom.xml` | modified | Added `service-api`, `spring-boot-starter-web`, `spring-boot-starter-tomcat` (provided). Added `spring-boot-maven-plugin` for WAR packaging |
| `ChangeLogsApplication.java` | new | Spring Boot entry point. Extends `SpringBootServletInitializer` for WAR deployment |
| `ChangeLogsController.java` | new | `@RestController` at `/api/changelogs`. Four REST endpoints replacing the Thrift service interface. Constructs a minimal Thrift `User` from the `X-User-Email` header for `assertUser()` validation |
| `ChangeLogsHandler.java` | modified | Changed from implementing Thrift `ChangeLogsService.Iface` to a plain `@Service`. Public API uses POJO types; internally uses `ThriftConverter` to communicate with `ChangeLogsDatabaseHandler`. Wraps all Thrift exceptions and converts them to POJO `SW360Exception` |
| `ChangeLogsServlet.java` | deleted | Was the Thrift servlet wrapper. No longer needed |
| `src/main/webapp/WEB-INF/web.xml` | deleted | Was the Thrift servlet registration. Spring Boot auto-configures |
| `src/main/webapp/index.jsp` | deleted | Placeholder JSP, no longer needed |
| `application.yml` | new | Spring Boot config: `spring.application.name: changelogs` |

REST Endpoints:

| Method | Path | Purpose |
|--------|------|---------|
| `GET` | `/api/changelogs/{changeLogId}` | Get a single changelog by its ID |
| `DELETE` | `/api/changelogs/{docId}` | Delete all changelogs for a document (requires `X-User-Email` header) |
| `GET` | `/api/changelogs/doc/{docId}` | Get all changelogs for a document (requires `X-User-Email` header) |
| `GET` | `/api/changelogs/doc/{docId}/page` | Get paginated changelogs for a document (requires `X-User-Email` header, accepts `@ModelAttribute PaginationData` via query params) |

### Layer 4 — REST Client (rest/resource-server)

| File | Action | Purpose |
|------|--------|---------|
| `Sw360ChangeLogService.java` | modified | Replaced `ThriftClients.makeChangeLogsClient()` with `RestClient` HTTP calls to backend endpoints. Sends `X-User-Email` header. Deserializes `PaginatedResult<ChangeLogs>` and reconstructs `Map<PaginationData, List<ChangeLogs>>` to maintain the existing contract with `ChangeLogController` |
| `ChangeLogController.java` | modified | Replaced Thrift `PaginationData` and `ChangeLogs` imports with POJO versions. Replaced `TException` with `Exception`. Constructs `PaginationResult` directly with null comparator to bypass the Thrift-typed `ResourceComparatorGenerator` (see "Design Decisions" below). Removed unused imports (`SW360Constants`, `URISyntaxException`, `ResourceClassNotFoundException`, `PaginationParameterException`) |

### Layer 5 — Infrastructure Fixes

| File | Action | Purpose |
|------|--------|---------|
| `PaginationResult.java` | modified | Made 3-arg constructor `public` (was package-private). Required so `ChangeLogController` can construct pre-paginated results directly, bypassing the Thrift-typed comparator pipeline |

### Layer 6 — Tests

| File | Action | Purpose |
|------|--------|---------|
| `ChangeLogTest.java` | modified | Replaced all Thrift imports with POJO versions. Changed `TException` → `Exception` in mock setup. Fixed `setTotalRowCount(2)` → `setTotalRowCount(2L)` for POJO `Long` type compatibility. Removed unused `Pageable` import |
| `ChangeLogSpecTest.java` | modified | Replaced Thrift imports with POJO versions. Changed `throws TException, IOException` → `throws Exception`. Fixed `setTotalRowCount(2)` → `setTotalRowCount(2L)`. Removed unused `IOException` import |

### Layer 7 — Incidental Changes

| File | Action | Purpose |
|------|--------|---------|
| `backend/configurations/src/main/resources/application.yml` | modified | Added EPL-2.0 license header |
| `backend/health/src/main/resources/application.yml` | modified | Added EPL-2.0 license header |
| `backend/health/HealthApplication.java` | modified | Whitespace fix in license comment block |
| `backend/health/db/HealthDatabaseHandler.java` | modified | Trailing whitespace removed from import line |
| `backend/search/src/main/resources/application.yml` | modified | Added EPL-2.0 license header |

### Design Decisions

**1. User Resolution from `X-User-Email` Header**

The `ChangeLogsDatabaseHandler` accepts a Thrift `User` parameter, and the handler calls `assertUser(user)` which validates that `email` and `department` are non-empty. Since the database handler doesn't actually use the User for permission checks (it's only passed through), the controller constructs a minimal Thrift `User` with the email from the header and `department = "REST"` as a lightweight placeholder that satisfies the assertion.

**2. `PaginatedResult<T>` Wrapper**

In the old Thrift architecture, `Map<PaginationData, List<T>>` was used extensively but only in-memory — it was never serialized to JSON. With REST, this map goes over HTTP, and Jackson serializes map keys by calling `toString()`, producing an opaque string like `PaginationData(ascending=true, ...)` that can't be deserialized back. The `PaginatedResult<T>` wrapper solves this by giving both the pagination metadata and the data list their own named JSON fields. The resource-server client deserializes the wrapper and reconstructs the `Map` to maintain backward compatibility with `ChangeLogController`.

**3. Direct `PaginationResult` Construction (Runtime ClassCastException)**

The resource-server's `RestControllerHelper.paginationResultFromPaginatedList()` internally uses `ResourceComparatorGenerator` which creates `Comparator<ThriftChangeLogs>`. When applied to a list of POJO `ChangeLogs`, this throws `ClassCastException`. Since the backend already returns data sorted and paginated, the controller constructs `PaginationResult` directly with a null comparator, bypassing the Thrift-bound comparator pipeline entirely. This required making `PaginationResult`'s 3-arg constructor public.

**4. PaginationData as Individual Query Parameters**

The resource-server client sends pagination fields as individual query params (`?ascending=false&displayStart=0&rowsPerPage=10&sortColumnNumber=0`) instead of serializing the entire `PaginationData` object as a single param. The backend's `@ModelAttribute PaginationData` automatically binds these individual params into the POJO.

### How Everything Connects

```
Browser/Client
    │
    ▼
resource-server (resource.war)
    │  RestClient → GET http://localhost:8080/changelogs/api/changelogs/doc/{id}/page
    │               ?ascending=false&displayStart=0&rowsPerPage=20&sortColumnNumber=0
    │               Header: X-User-Email: admin@sw360.org
    │  Sw360ChangeLogService deserializes PaginatedResult<ChangeLogs>
    │  Reconstructs Map<PaginationData, List<ChangeLogs>> for ChangeLogController
    │  ChangeLogController builds PaginationResult directly (null comparator)
    │  Wraps into HATEOAS PagedModel with pagination links
    ▼
backend/changelogs (changelogs.war)
    │  ChangeLogsController receives @ModelAttribute PaginationData
    │  Constructs Thrift User from X-User-Email header
    │  ChangeLogsHandler uses ThriftConverter to bridge POJOs ↔ Thrift
    │  Calls ChangeLogsDatabaseHandler (still Thrift-based)
    │  Returns PaginatedResult<ChangeLogs> (wrapper DTO)
    ▼
CouchDB (sw360changelogs)
```

### Verification

- All 5 curl test cases return identical responses before and after migration (with expected addition of `revision` and `dbName` fields previously filtered by Thrift Jackson mixin)
- Pagination links (`first`, `next`, `last`) and page metadata (`size`, `totalElements`, `totalPages`, `number`) are correct
- Empty-changelog and nonexistent-document edge cases behave identically
- All 6 changelog tests pass (5 integration + 1 restdocs spec)
- Full project compilation: BUILD SUCCESS across all 40+ modules

### CODE REVIEW

**Comment by Shivamrut** [comment]
Adding sequence number for order of migration: service-4

File: .vscode/settings.json
Comment (GMishx) [inline .vscode/settings.json:1]: Again :-)

File: backend/changelogs/src/main/java/org/eclipse/sw360/changelogs/ChangeLogsApplication.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
+/*
+ *  Copyright Shivamrut<gshivamrut@gmail.com>, 2026. Part of the SW360 Portal Project.
+ * 
+ *  This program and the accompanying materials are made
+ *  available under the terms of the Eclipse Public License 2.0
+ *  which is available at https://www.eclipse.org/legal/epl-2.0/
+ * 
+ *  SPDX-License-Identifier: EPL-2.0
+*/
```
Comment (GMishx) [inline backend/changelogs/src/main/java/org/eclipse/sw360/changelogs/ChangeLogsApplication.java]: Please fix formatting.

File: backend/common/src/main/java/org/eclipse/sw360/common/converter/ThriftConverter.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
+/*
+ *  Copyright Shivamrut<gshivamrut@gmail.com>, 2026. Part of the SW360 Portal Project.
+ * 
+ *  This program and the accompanying materials are made
+ *  available under the terms of the Eclipse Public License 2.0
+ *  which is available at https://www.eclipse.org/legal/epl-2.0/
+ * 
+ *  SPDX-License-Identifier: EPL-2.0
+*/
```
Comment (GMishx) [inline backend/common/src/main/java/org/eclipse/sw360/common/converter/ThriftConverter.java]: Here as well.

File: backend/configurations/src/main/java/org/eclipse/sw360/configurations/ConfigurationsApplication.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
+/*
+ *  Copyright Shivamrut<gshivamrut@gmail.com>, 2026. Part of the SW360 Portal Project.
+ * 
+ *  This program and the accompanying materials are made
+ *  available under the terms of the Eclipse Public License 2.0
+ *  which is available at https://www.eclipse.org/legal/epl-2.0/
+ * 
+ *  SPDX-License-Identifier: EPL-2.0
+*/
```
Comment (GMishx) [inline backend/configurations/src/main/java/org/eclipse/sw360/configurations/ConfigurationsApplication.java]: Here as well.

File: backend/configurations/src/main/java/org/eclipse/sw360/configurations/ConfigurationsController.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
+/*
+ *  Copyright Shivamrut<gshivamrut@gmail.com>, 2026. Part of the SW360 Portal Project.
+ * 
+ *  This program and the accompanying materials are made
+ *  available under the terms of the Eclipse Public License 2.0
+ *  which is available at https://www.eclipse.org/legal/epl-2.0/
+ * 
+ *  SPDX-License-Identifier: EPL-2.0
+*/
```
Comment (GMishx) [inline backend/configurations/src/main/java/org/eclipse/sw360/configurations/ConfigurationsController.java]: Here as well.

File: libraries/service-api/src/main/java/org/eclipse/sw360/datahandler/services/changelogs/ChangedFields.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
+/*
+ *  Copyright Shivamrut<gshivamrut@gmail.com>, 2026. Part of the SW360 Portal Project.
+ * 
+ *  This program and the accompanying materials are made
+ *  available under the terms of the Eclipse Public License 2.0
+ *  which is available at https://www.eclipse.org/legal/epl-2.0/
+ * 
+ *  SPDX-License-Identifier: EPL-2.0
+*/
```
Comment (GMishx) [inline libraries/service-api/src/main/java/org/eclipse/sw360/datahandler/services/changelogs/ChangedFields.java]: Here as well.

**Review by GMishx — CHANGES_REQUESTED** [review]

**Comment by Shivamrut** [comment]
Hi @GMishx Please let me rebase the PR after last PR merge. Since each new service is checked out wrt previous service so these formatting fixes will automatically be applied once I rebase it.

**Comment by Shivamrut** [comment]
Hi @GMishx fixed the formatting and merge conflicts. Please review

**Review by GMishx — APPROVED** [review]
LGTM
